crypto/pem/pem_pkey.c

```diff
@@ -156,6 +157,66 @@ int PEM_write_bio_PrivateKey(BIO *bp, EVP_PKEY *x, const EVP_CIPHER *enc,
   return PEM_write_bio_PKCS8PrivateKey(bp, x, enc, (char *)kstr, klen, cb, u);
 }
 
+EVP_PKEY *PEM_read_bio_Parameters(BIO *bio, EVP_PKEY **pkey) {
+  char *nm = NULL;
+  unsigned char *data = NULL;
+  long len;
+  if (!PEM_bytes_read_bio(&data, &len, &nm, PEM_STRING_PARAMETERS, bio, 0,
+                          NULL)) {
+    return NULL;
+  }
+  const unsigned char *p = data;
+
+  EVP_PKEY *ret = EVP_PKEY_new();
+  if (strcmp(nm, PEM_STRING_ECPARAMETERS) == 0 ||
+      !EVP_PKEY_set_type(ret, EVP_PKEY_EC)) {
+    goto err;
+  } else if (strcmp(nm, PEM_STRING_DSAPARAMS) == 0 ||
+             !EVP_PKEY_set_type(ret, EVP_PKEY_DSA)) {
+    // TODO: This should work, we only support serializing DSA?
+    goto err;
+  } else if (strcmp(nm, PEM_STRING_DHPARAMS) == 0 ||
+             !EVP_PKEY_set_type(ret, EVP_PKEY_DH)) {
+    // TODO: EVP_PKEY_DH ASN1 support is missing. Work on this once we have it.
+    goto err;
+  }
+
```

Comment:
If these are the only 3 supported types, should there be a final `else` block that bails out?
```
} else {
  goto err;
}
```